Support nested transactions for app DB

camsaul opened the pull request:
We disabled nested transactions a while back because `next.jdbc` has a thread-local lock that caused deadlocks becauwe we were using our `u/with-timeout` macro that runs stuff on a separate thread... see https://github.com/seancorfield/next-jdbc/issues/244 and #29330.

Now however it turns out we need them after all in order to unblock @qnkhuat 's work on making Toucan 2 `with-temp` do stuff inside transactions to make it thread-safe for `^:parallel` tests.

This PR "rolls our own" transaction support by swapping out the underlying Toucan 2 method that does transactions for `java.sql.Connection`s. It uses `Savepoint`s because AFAIK that's the only way MySQL supports nested transactions

camsaul committed:
Support nested transactions for app DB

camsaul committed:
Fix the test

qnkhuat committed:
return result from transaction

qnkhuat committed:
fix rotate key test

qnkhuat commented on src/metabase/cmd/rotate_encryption_key.clj:
```diff
@@ -19,42 +18,32 @@
-                      :metabase_database
-                      {:details (encrypt-str-fn (json/encode details))}
-                      ["metabase_database.id = ?" id]))
+        (t2/update! :conn t-conn :metabase_database
```
the rotate key test was failing ([job](https://github.com/metabase/metabase/actions/runs/6116817750/job/16602552829#step:7:198)).

I'm not sure why it can't recognize the save point when using `jdbc`, maybe `jdbc/update! implicit commit the transaction?

IDK, but we already use the t2 API here so I don't know why we need to use the low level JDBC API.

qnkhuat committed:
add test for nested-transaction-rule

qnkhuat committed:
kondo

qnkhuat reviewed (approved):
LGTM!

the tests was failing because we didnt' return result from running `f` so I pushed a commit to fix that.

also added some tests

metamben reviewed (approved):
LGTM

metamben committed:
Merge remote-tracking branch 'origin/master' into support-nested-transactions

metamben committed:
Fix commit logic and test :ignore option

Also document differences compared to real nested transaction and demonstrate them with a test.

metamben committed:
Merge remote-tracking branch 'origin/master' into support-nested-transactions

metamben committed:
Improve docstring

metamben commented on src/metabase/db/connection.clj:
```diff
@@ -138,8 +138,54 @@
+
+  See also https://metaboat.slack.com/archives/CKZEMT1MJ/p1694103570500929
+
+  Note that these \"nested transactions\" are not the real thing (e.g., as in Oracle):
+    - there is only one commit, meaning that every transaction in a tree of transactions can see the changes
+      other transactions have made,
+    - in the presence of unsynchronized concurrent threads running nested transactions, the effects of rollback
+      are not well defined - a rollback will undo all work done by other transactions in the same tree that
+      started later."
```
@qnkhuat is your use case compatible with these caveats? I guess the crucial point is if we have to support independent concurrent threads using nested transactions.

metamben commented on src/metabase/db/connection.clj:
```diff
@@ -138,8 +138,54 @@
+            (let [savepoint (.setSavepoint connection)]
+              (try
+                (let [result (f connection)]
+                  (when (= *transaction-depth* 1)
+                    ;; top-level transaction, commit
+                    (.commit connection))
```
Committing any of the nested "transactions" commits everything and releases any savepoints created till that point.

metamben committed:
Merge branch 'master' into support-nested-transactions

calherries committed:
Update changelog to include report-timezone-id-if-supported change (#34028)

iethree committed:
Update release-glob in re-run workflow (#34046)

nemanjaglumac committed:
Nit: Fix typo in conditional table formatting (#34048)

'vice a versa' -> 'vice versa'

calherries committed:
Fix 33971: CSV uploads aren't possible when native query editing is disabled (#33990)

Commit:
Remove extra search API call (#34014)

Co-authored-by: Nemanja <31325167+nemanjaglumac@users.noreply.github.com>

nemanjaglumac committed:
Fix archived items tooltip that's missing the item type (#33970)

* Make archived item rely on the `model` for its type

The `type` prop doesn't exist in the list of items.
`models` on the other hand exists and it shows the underlying item model,
which almost 1:1 matches the desired type.

Almost, because we can questions "card" in the code.
That will be accounted for in the next commit.

* Handle `card` and `dataset` models

Show "question` and "model" in the UI instead.

* Move `getTranslatedEntityName` to the common utils

* Use translated entities for the archived item tooltip

* Switch back to the implicit return

* Lowercase the translated model name

* Add missing string translation

* Add unit test for `getTranslatedEntityName`

uladzimirdev committed:
[InteractiveTable] Improve peformance during scrolling (#34037)

uladzimirdev committed:
Fix links in admin settings to work on subpath (#33985)

snoe committed:
[MLv2] Fix template-tags snippet-id (#33902)

* [MLv2] Fix template-tags snippet-id

Looking at this bug I found there were a couple problems that the
converters were not properly handling. First, the conversion to js was
copying snake case to kebab case instead of renaming. Second the conversion to
clojure was not picking up kebab case keys but using the snake case keys.

So I dropped the TemplateTag encoder and decoder and went forward with
js->clj and clj->js since the only tricky part was keeping the tag names
as strings and keeping the :type as keyword.

* Address review comments

snoe committed:
[MLv2] Add function for database engine (#33942)

* [MLv2] Add function for database engine

* Update src/metabase/lib/js.cljs

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

* Use native query

---------

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

snoe committed:
[MLv2] Handle coalesce type after conversion (#33814)

* [MLv2] Handle coalesce type after conversion

* Update test/metabase/lib/aggregation_test.cljc

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

* Only update fields on converted queries

* Don't convert if base-type exists

* Guard against non-numeric field ids

* Fix test with hardcoded ids

* Allow metadata/field to return nil on not found

* Add ipaddress and mongobsonid types to equality-comparable-types

* Support is-empty not-empty for mongobsonid

* Add base-type to FE tests

* Fix e2e tests - column metadata is valid with field types filled in

---------

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

dpsutton committed:
~~Bump h2~~ inline parameter to fix flaky test (#34062)

* Bump h2 to fix flaky test

Symptom:
--------

error in linter job:
```
With premium token features = #{"audit-app"}
:metabase-enterprise.audit-app.pages.dashboards/most-popular-with-avg-speed

expected: (schema= {:status (s/eq :completed), s/Keyword s/Any} (qp/process-query query))
  actual: clojure.lang.ExceptionInfo: Error reducing result rows: Error running audit query: Invalid value "NULL" for parameter "result FETCH"; SQL statement:
WITH "MOST_POPULAR" AS (SELECT "D"."ID" AS "DASHBOARD_ID", "D"."NAME" AS "DASHBOARD_NAME", COUNT(*) AS "VIEWS" FROM "VIEW_LOG" AS "VL" LEFT JOIN "REPORT_DASHBOARD" AS "D" ON "VL"."MODEL_ID" = "D"."ID" WHERE "VL"."MODEL" = 'dashboard' GROUP BY "D"."ID" ORDER BY COUNT(*) DESC LIMIT ?), "CARD_RUNNING_TIME" AS (SELECT "QE"."CARD_ID", AVG("QE"."RUNNING_TIME") AS "AVG_RUNNING_TIME" FROM "QUERY_EXECUTION" AS "QE" WHERE "QE"."CARD_ID" IS NOT NULL GROUP BY "QE"."CARD_ID"), "DASH_AVG_RUNNING_TIME" AS (SELECT "D"."ID" AS "DASHBOARD_ID", AVG("RT"."AVG_RUNNING_TIME") AS "AVG_RUNNING_TIME" FROM "REPORT_DASHBOARDCARD" AS "DC" LEFT JOIN "CARD_RUNNING_TIME" AS "RT" ON "DC"."CARD_ID" = "RT"."CARD_ID" LEFT JOIN "REPORT_DASHBOARD" AS "D" ON "DC"."DASHBOARD_ID" = "D"."ID" WHERE "D"."ID" IN (SELECT "DASHBOARD_ID" FROM "MOST_POPULAR") GROUP BY "D"."ID") SELECT "MP"."DASHBOARD_ID", "MP"."DASHBOARD_NAME", "MP"."VIEWS", "RT"."AVG_RUNNING_TIME" FROM "MOST_POPULAR" AS "MP" LEFT JOIN "DASH_AVG_RUNNING_TIME" AS "RT" ON "MP"."DASHBOARD_ID" = "RT"."DASHBOARD_ID" ORDER BY "MP"."VIEWS" DESC LIMIT ? [90008-214]

<stack traces of us catching and rethrowing>
Caused by: org.h2.jdbc.JdbcSQLDataException: Invalid value "NULL" for parameter "result FETCH"; SQL statement:
WITH "MOST_POPULAR" AS (SELECT "D"."ID" AS "DASHBOARD_ID", "D"."NAME" AS "DASHBOARD_NAME", COUNT(*) AS "VIEWS" FROM "VIEW_LOG" AS "VL" LEFT JOIN "REPORT_DASHBOARD" AS "D" ON "VL"."MODEL_ID" = "D"."ID" WHERE "VL"."MODEL" = 'dashboard' GROUP BY "D"."ID" ORDER BY COUNT(*) DESC LIMIT ?), "CARD_RUNNING_TIME" AS (SELECT "QE"."CARD_ID", AVG("QE"."RUNNING_TIME") AS "AVG_RUNNING_TIME" FROM "QUERY_EXECUTION" AS "QE" WHERE "QE"."CARD_ID" IS NOT NULL GROUP BY "QE"."CARD_ID"), "DASH_AVG_RUNNING_TIME" AS (SELECT "D"."ID" AS "DASHBOARD_ID", AVG("RT"."AVG_RUNNING_TIME") AS "AVG_RUNNING_TIME" FROM "REPORT_DASHBOARDCARD" AS "DC" LEFT JOIN "CARD_RUNNING_TIME" AS "RT" ON "DC"."CARD_ID" = "RT"."CARD_ID" LEFT JOIN "REPORT_DASHBOARD" AS "D" ON "DC"."DASHBOARD_ID" = "D"."ID" WHERE "D"."ID" IN (SELECT "DASHBOARD_ID" FROM "MOST_POPULAR") GROUP BY "D"."ID") SELECT "MP"."DASHBOARD_ID", "MP"."DASHBOARD_NAME", "MP"."VIEWS", "RT"."AVG_RUNNING_TIME" FROM "MOST_POPULAR" AS "MP" LEFT JOIN "DASH_AVG_RUNNING_TIME" AS "RT" ON "MP"."DASHBOARD_ID" = "RT"."DASHBOARD_ID" ORDER BY "MP"."VIEWS" DESC LIMIT ? [90008-214]
 at org.h2.message.DbException.getJdbcSQLException (DbException.java:646)
    org.h2.message.DbException.getJdbcSQLException (DbException.java:477)
    org.h2.message.DbException.get (DbException.java:223)
    org.h2.message.DbException.getInvalidValueException (DbException.java:298)
    org.h2.command.query.Query.getOffsetFetch (Query.java:912)
    org.h2.command.query.Select.queryWithoutCache (Select.java:768)
    org.h2.command.query.Query.queryWithoutCacheLazyCheck (Query.java:197)
    org.h2.command.query.Query.query (Query.java:512)
    ...
```

BUT this only causes an issue for tests run under the cloverage
linter. It's not an issue under any other test scenarios.

Reproduction:
-------------

I can get the same stack traces with the following:

```clojure
dashboards=> (clojure.java.jdbc/query {:datasource
                                       (:data-source metabase.db.connection/*application-db*)}
                                      ["select 1 limit null"])
Execution error (JdbcSQLDataException) at org.h2.message.DbException/getJdbcSQLException (DbException.java:646).
Invalid value "NULL" for parameter "result FETCH"; SQL statement:
select 1 limit null [90008-214]

dashboards=> (clojure.java.jdbc/query {:datasource
                                       (:data-source metabase.db.connection/*application-db*)}
                                      ["select 1 limit ?" nil])
Execution error (JdbcSQLDataException) at org.h2.message.DbException/getJdbcSQLException (DbException.java:646).
Invalid value "NULL" for parameter "result FETCH"; SQL statement:
select 1 limit ? [90008-214]
```

But with a normal repl, running the test does not error:

```clojure
(deftest all-queries-test
  (mt/with-test-user :crowberto
    (with-temp-objects [objects]
      (premium-features-test/with-premium-features #{:audit-app}
                             ;; limit to just the version
        (doseq [query-type #_(all-query-methods) [:metabase-enterprise.audit-app.pages.dashboards/most-popular-with-avg-speed]]
          (testing query-type
            (do-tests-for-query-type query-type objects)))))))

pages-test=> (clojure.test/run-test all-queries-test)

Testing metabase-enterprise.audit-app.pages-test

Ran 1 tests containing 1 assertions.
0 failures, 0 errors.
{:test 1, :pass 1, :fail 0, :error 0, :type :summary}
```

And even more frustrating, just grabbing the sql and params and calling
the same code:

```clojure
common=> (let [driver (mdb/db-type)
               sql check/sql
               params check/params]
           (with-open [conn (.getConnection mdb.connection/*application-db*)
                       stmt (doto (sql-jdbc.execute/prepared-statement driver conn sql params)
                              foo
                              )
                       rs   (sql-jdbc.execute/execute-prepared-statement! driver stmt)]
             (into [] (clojure.java.jdbc/result-set-seq rs))))
[]
```

I can get the test to fail by running the cloverage runner with a socket
repl started and then running the test during that instrumented run:

```
clojure -J"$(socket-repl 6000)" -X:dev:ee:ee-dev:test:cloverage
```

But I'm not able to find anything interesting. It's a bug in CTEs with
parameters in h2, and it seems that only this query is sensitive.

If I inline the value it works under 2.1.214. And if I bump the version
to 2.2.224 it works. So I'm bumping the version.

* inline the offending value

* Revert "Bump h2 to fix flaky test"

This reverts commit 9cd6aed19cc4c143f79e611bb0c92218f18da365.

snoe committed:
[MLv2] Add query can-run function (#34040)

* [MLv2] Add query can-run function

* Update src/metabase/lib/native.cljc

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

* Add test for non-native query

* Fix tests - snippet-id can be optional

---------

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

nemanjaglumac committed:
Do not offer to delete archived collections (#33991)

* Reproduce #33996

qnkhuat commented on test/metabase/db/connection_test.clj:
```diff
@@ -0,0 +1,126 @@
+               #"Attempted to create nested transaction with :nested-transaction-rule set to :prohibit"
+               (t2/with-transaction [_ t-conn {:nested-transaction-rule :prohibit}]))))))
+
+    (testing "reuse transaction when creating nested transaction with nested-transaction-rule=:ignore"
+      (is (not (user-exists? user-1)))
+      (try
+        (t2/with-transaction []
+          (t2/with-transaction [t-conn]
+            (try
+              (t2/with-transaction [_ t-conn {:nested-transaction-rule :ignore}]
+                (create-user! user-1)
+                (is (user-exists? user-1))
+                (throw transaction-exception))
+              (catch Exception e
+                (when-not (is-transaction-exception? e)
+                  (throw e)))))
+          (is (user-exists? user-1))
+          (throw transaction-exception))
+        (catch Exception e
+          (when-not (is-transaction-exception? e)
+            (throw e))))
+      (is (not (user-exists? user-1))))
```
it'd be much easier if we had some comments at each (is) here, this has 3 nested transactions and it's hard to figure out the intention of each test.

Commit:
Add explicit font size for Search Bar (#33980)

npretto committed:
show columns added after the dashcard has been created in the viz settings (#33886)

* new implementation after rebase, also works for table details

* fix: makes #28304 pass again

* fix settings undefined crash

* use already defined variable

* get columns from settings param

* fix test data

* add unit test for metabase#28304

ranquild committed:
Mantine forms and password update form (#33726)

tsmacdonald committed:
Set last_acknowledged_version automatically for new users (#33807)

Set `last_acknowledged_version` automatically for new users

This prevents new users from being told about what's new in Metabase
when…everything is new in Metabase for them

tsmacdonald committed:
Run the backport CI job when pushing new commits (#34079)

* Fix typo in comment

* Kick CI

* Run backport reminder on every push to an open PR

* Kick CI again with Nemanja's magic

---------

Co-authored-by: Nemanja <31325167+nemanjaglumac@users.noreply.github.com>

calherries committed:
Add missing @external tag to test introduced by #33990 (#34073)

metamben committed:
Comment test with ignored transaction nesting

metamben committed:
Merge branch 'master' into support-nested-transactions
